### PR TITLE
Fixes possible serialization problems putting object into the session…

### PR DESCRIPTION
**Pull request by fmartelli**
Serialize XML objects before putting them into the session store

### CODE REVIEW

File: pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java [hunk 1]
```diff
@@ -34,7 +37,7 @@
     /**
      * Internal store for messages, corresponding to the object in session.
      */
-    private LinkedHashMap<String, XMLObject> internalMessages;
+    private LinkedHashMap<String, String> internalMessages;
```
Comment (mmoayyed) [inline pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java]: I realize this has always been here, but is there a reason LinkedHashMap is required? Could we just use the usual HashMap instead?

File: pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java [hunk 2]
```diff
@@ -86,17 +89,27 @@ public void set(final String messageID, final XMLObject message) {
...
     @Override
     public Optional<XMLObject> get(final String messageID) {
-        final LinkedHashMap<String, XMLObject> messages = getMessages();
-        final XMLObject o = messages.get(messageID);
+        final LinkedHashMap<String, String> messages = getMessages();
+        final String o = messages.get(messageID);
         if (o == null) {
-            log.debug("Message {} not found in session {}", messageID, context.getSessionStore().getOrCreateSessionId(context));
+            log.debug("Message {} not found in session {}",
+                    messageID, context.getSessionStore().getOrCreateSessionId(context));
```
Comment (mmoayyed) [inline pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java]: Not sure you need to log `context.getSessionStore().getOrCreateSessionId(context)`, as that might actually create a session id? @leleuj would know best.

File: pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java [hunk 3]
```diff
@@ -86,17 +89,27 @@ public void set(final String messageID, final XMLObject message) {
...
-        return Optional.of(o);
+
+        try {
+            XMLObject message = XMLObjectSupport.unmarshallFromReader(
+                    Configuration.getParserPool(), new StringReader(o));
+            return Optional.of(message);
+        } catch (Exception e) {
+            log.error("Error unmarshalling message from input stream", e);
+            return Optional.empty();
+        }
```
Comment (mmoayyed) [inline pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java]: You should likely move these lines into a method next to one in Configuration that does the serialization.

File: pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java [hunk 4]
```diff
@@ -86,17 +89,27 @@ public void set(final String messageID, final XMLObject message) {
...
-        log.debug("Message {} found in session {}, clearing", messageID, context.getSessionStore().getOrCreateSessionId(context));
+        log.debug("Message {} found in session {}, clearing",
+                messageID, context.getSessionStore().getOrCreateSessionId(context));
         messages.clear();
         updateSession(messages);
-        return Optional.of(o);
+
+        try {
+            XMLObject message = XMLObjectSupport.unmarshallFromReader(
+                    Configuration.getParserPool(), new StringReader(o));
```
Comment (mmoayyed) [inline pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java]: Is the StringReader expected to be closed by the caller(pac4j)?

File: pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java (hunk 1 above)
Comment (fmartelli) [reply pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java]: Good question!

File: pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java (hunk 2 above)
Comment (fmartelli) [reply pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java]: I left the log statement as it was.
Please let me know if you prefer to change that line

Comment (leleuj) [reply pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java]: Indeed, the `getOrCreateSessionId` method could create a session if it does not exist.
Given it's in the webflow of the SAML login process, it's unlikely though.
I think we can keep these for now.

File: pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java (hunk 1 above)
Comment (leleuj) [reply pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java]: I guess that there might be some performance idea behind that: messages are saved and should be retrieved in order...

Comment (fmartelli) [reply pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java]: thanks

File: pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java (hunk 2 above)
Comment (fmartelli) [reply pac4j-saml/src/main/java/org/pac4j/saml/store/HttpSessionStore.java]: thanks

**Pull request merged**